### PR TITLE
feature/GEODE-3571

### PR DESCRIPTION
**Pull request by mhansonp**
GEODE-3571: Move to the Builder and Fluent models.

### CODE REVIEW

File: clicache/integration-test/AttributesFactoryTestsN.cs [hunk 1]
```diff
@@ -15,94 +15,29 @@
...
+  [Category("unicast_only")]
+  [Category("generics")]
+  public class AttributesFactoryTests : UnitTests
+  {
+    [Test]
+    public void fluentModeltest()
+    {
+      AttributesFactory<TKey, TValue> af = new AttributesFactory<TKey, TValue>();
+      Apache.Geode.Client.RegionAttributes<TKey, TValue> rattrs = af.SetLruEntriesLimit(0).SetInitialCapacity(5).CreateRegionAttributes();
+      Assert.IsNotNull(rattrs);
```
Comment (pivotal-jbarrett) [inline clicache/integration-test/AttributesFactoryTestsN.cs]: I would suggest adding a bit more to this test. If the point is to assert that the return from each set call is the same object you called the setter an then I would put that in the test. All this test does is show that the last call creates a region attribute. The intermediate set calls could be returning different factory objects.

File: clicache/src/geode_defs.hpp [hunk 2]
```diff
@@ -33,6 +33,11 @@
 // Disable native code generation warning
 #pragma warning(disable: 4793)
 
+// Disable Native Conditional Unique warning
+#pragma warning(disable: 4383)
+
+// C4103alignment changed after including header, may be due to missing #pragma pack(pop)
+#pragma warning(disable: 4103)
```
Comment (pivotal-jbarrett) [inline clicache/src/geode_defs.hpp]: Are these changes necessary to relevant to the immediate issue being addressed?

File: clicache/src/impl/CacheWriter.hpp [hunk 3]
```diff
@@ -76,8 +76,7 @@ namespace Apache
             RegionEvent<TKey, TValue> gevent(ev->GetNative());
             return m_writer->BeforeRegionDestroy(%gevent);
           }
-          
-          virtual void Close(Apache::Geode::Client::Region<Object^, Object^>^ region) override
+          virtual void Close(Apache::Geode::Client::Region<Object^, Object^>^ region)
```
Comment (pivotal-jbarrett) [inline clicache/src/impl/CacheWriter.hpp]: Why did the override get dropped here?

File: cppcache/test/AttributesFactoryTest.cpp [hunk 4]
```diff
@@ -23,36 +23,40 @@ using namespace apache::geode::client;
 
 TEST(AttributesFactoryTest, setEntryIdleTimeoutSeconds) {
   AttributesFactory attributesFactory;
-  attributesFactory.setEntryIdleTimeout(ExpirationAction::DESTROY,
-                                        std::chrono::seconds(10));
   std::shared_ptr<RegionAttributes> regionAttributes =
-      attributesFactory.createRegionAttributes();
+      attributesFactory
```
Comment (pivotal-jbarrett) [inline cppcache/test/AttributesFactoryTest.cpp]: Similar to the comments above I would add a test that makes sure the the return from each set call is a reference to the same object we called the set on.

File: cppcache/include/geode/Exception.hpp [hunk 5]
```diff
@@ -19,6 +19,7 @@
 
 #ifndef GEODE_EXCEPTION_H_
 #define GEODE_EXCEPTION_H_
+#pragma warning(disable : 4275)
```
Comment (pivotal-jbarrett) [inline cppcache/include/geode/Exception.hpp]: What does this disable? Comment?

File: cppcache/src/AttributesFactory.cpp [hunk 6]
```diff
@@ -38,92 +38,100 @@ AttributesFactory::AttributesFactory(
...
-
 std::unique_ptr<RegionAttributes> AttributesFactory::createRegionAttributes() {
   std::shared_ptr<RegionAttributes> res;
   validateAttributes(m_regionAttributes);
   return std::unique_ptr<RegionAttributes>(
       new RegionAttributes(m_regionAttributes));
 }
 
-void AttributesFactory::validateAttributes(RegionAttributes& attrs) {
+void  AttributesFactory::validateAttributes(RegionAttributes& attrs) {
```
Comment (pivotal-jbarrett) [inline cppcache/src/AttributesFactory.cpp]: Formatting

File: cppcache/src/AttributesFactory.cpp [hunk 7]
```diff
@@ -164,44 +172,52 @@ void AttributesFactory::validateAttributes(RegionAttributes& attrs) {
   }
 }
 
-void AttributesFactory::setLruEntriesLimit(const uint32_t entriesLimit) {
+AttributesFactory&  AttributesFactory::setLruEntriesLimit(const uint32_t entriesLimit) {
```
Comment (pivotal-jbarrett) [inline cppcache/src/AttributesFactory.cpp]: Formatting

Comment (pivotal-jbarrett) [reply cppcache/src/AttributesFactory.cpp]: Looks like all these have extra spaces between the return type and method name.

File: cppcache/src/PoolFactory.cpp [hunk 8]
```diff
@@ -71,17 +71,18 @@ PoolFactory::PoolFactory(const Cache& cache)
 
 PoolFactory::~PoolFactory() {}
 
-void PoolFactory::setFreeConnectionTimeout(
+PoolFactory& PoolFactory::setFreeConnectionTimeout(
```
Comment (pivotal-jbarrett) [inline cppcache/src/PoolFactory.cpp]: Can we add a unit test for this class?

File: clicache/src/impl/CacheWriter.hpp (hunk 3 above)
Comment (mhansonp) [reply clicache/src/impl/CacheWriter.hpp]: it was very wrong in terms of matching the actual prototype, and I was removing a warning. I actually didn't intend to check this in.

Comment (mhansonp) [reply clicache/src/impl/CacheWriter.hpp]: I know it is supposed to do that if someone in advertently changes the base function prototype, that is why I didn't intend to check that in.

File: cppcache/include/geode/Exception.hpp (hunk 5 above)
Comment (mhansonp) [reply cppcache/include/geode/Exception.hpp]: Hmm. I am surprised, I didn't comment.... I will fix that.

File: clicache/integration-test/AttributesFactoryTestsN.cs (hunk 1 above)
Comment (mhansonp) [reply clicache/integration-test/AttributesFactoryTestsN.cs]: Good point.

File: clicache/src/geode_defs.hpp (hunk 2 above)
Comment (mhansonp) [reply clicache/src/geode_defs.hpp]: These two warnings were causing over 1000 warnings, so I could not see the warnings I was generating. I took them out as they are CLI related, but not real issues.

File: clicache/integration-test/AttributesFactoryTestsN.cs [hunk 9]
```diff
@@ -24,20 +24,35 @@ namespace Apache.Geode.Client.UnitTests
...
   [TestFixture]
   [Category("group1")]
   [Category("unicast_only")]
   [Category("generics")]
   public class AttributesFactoryTests : UnitTests
   {
+
+    private UnitProcess m_client1, m_client2;
+
+    protected override ClientBase[] GetClients()
```
Comment (pivotal-jbarrett) [inline clicache/integration-test/AttributesFactoryTestsN.cs]: Why did this test require adding clients? As far as I can tell all the testing runs local to the test runner and are not invoked in remote clients.

File: cppcache/include/geode/Exception.hpp [hunk 10]
```diff
@@ -19,6 +19,12 @@
 
 #ifndef GEODE_EXCEPTION_H_
 #define GEODE_EXCEPTION_H_
+
+/*
+warning C4275:
+non dll-interface class 'std::exception' used as base for dll-interface
+class 'apache::geode::client::Exception'
+*/
 #pragma warning(disable : 4275)
```
Comment (pivotal-jbarrett) [inline cppcache/include/geode/Exception.hpp]: Pretty sure this should be pushed and then popped so that it only covers this particular class.
